dynamic/clientset/clientset.go

```diff
@@ -71,85 +70,82 @@ func (cs *Clientset) Kind(apiVersion, kind string) (*ResourceClient, error) {
-	resource *dynamicdiscovery.APIResource
-}
-
-func (rc *ResourceClient) Namespace(namespace string) *NamespacedResourceClient {
-	return &NamespacedResourceClient{
-		ResourceInterface: rc.NamespaceableResourceInterface.Namespace(namespace),
-		resource:          rc.resource,
-	}
-}
+	*dynamicdiscovery.APIResource
 
-func (rc *ResourceClient) Kind() string {
-	return rc.resource.Kind
+	rootClient dynamic.NamespaceableResourceInterface
 }
 
-func (rc *ResourceClient) GroupVersion() schema.GroupVersion {
-	return rc.resource.GroupVersion()
-}
-
-func (rc *ResourceClient) GroupResource() schema.GroupResource {
-	return schema.GroupResource{
-		Group:    rc.resource.GroupVersion().Group,
-		Resource: rc.resource.Name,
+// Namespace returns a copy of the ResourceClient with the client namespace set.
+//
+// This can be chained to set the namespace to something else.
+// Pass "" to return a client with the namespace cleared.
+// If the resource is cluster-scoped, this is a no-op.
+func (rc *ResourceClient) Namespace(namespace string) *ResourceClient {
+	// Ignore the namespace if the resource is cluster-scoped.
+	if !rc.Namespaced {
+		return rc
+	}
+	// Reset to cluster-scoped if provided namespace is empty.
+	ri := dynamic.ResourceInterface(rc.rootClient)
+	if namespace != "" {
+		ri = rc.rootClient.Namespace(namespace)
+	}
+	return &ResourceClient{
+		ResourceInterface: ri,
+		APIResource:       rc.APIResource,
+		rootClient:        rc.rootClient,
 	}
 }
 
-func (rc *ResourceClient) GroupVersionKind() schema.GroupVersionKind {
-	return rc.resource.GroupVersion().WithKind(rc.resource.Kind)
-}
-
-func (rc *ResourceClient) GroupVersionResource() schema.GroupVersionResource {
-	return rc.resource.GroupVersion().WithResource(rc.resource.Name)
-}
-
-func (rc *ResourceClient) APIResource() *dynamicdiscovery.APIResource {
-	return rc.resource
-}
-
-func (rc *ResourceClient) UpdateWithRetries(orig *unstructured.Unstructured, update func(obj *unstructured.Unstructured) bool) (result *unstructured.Unstructured, err error) {
-	return updateWithRetries(rc, rc.resource, orig, update)
-}
-
-func (nrc *NamespacedResourceClient) UpdateWithRetries(orig *unstructured.Unstructured, update func(obj *unstructured.Unstructured) bool) (result *unstructured.Unstructured, err error) {
-	return updateWithRetries(nrc, nrc.resource, orig, update)
-}
-
-func updateWithRetries(ri dynamic.ResourceInterface, resource *dynamicdiscovery.APIResource, orig *unstructured.Unstructured, update func(obj *unstructured.Unstructured) bool) (result *unstructured.Unstructured, err error) {
+// AtomicUpdate performs an atomic read-modify-write loop, retrying on
+// optimistic concurrency conflicts.
+//
+// It only uses the identity (name/namespace/uid) of the provided 'orig' object,
+// not the contents. The object passed to the update() func will be from a live
+// GET against the API server.
+//
+// This should only be used for unconditional writes, as in, "I want to make
```

Comment:
Did previous `UpdateWithRetries` usages mean to be unconditional writes?

Reply:
Yes, we should only have been using UpdateWithRetries in cases when we *don't* want the write to be conditional on the ResourceVersion being the same as our last read.

Reply:
Noted, thanks!